content/post/ddos-detection.md

```diff
@@ -0,0 +1,419 @@
+```
+
+As each server's status is computed independently of the others, we need a way to choose the partition for each incoming record. We could create a partition for each unique IP address, but that would require a partition index list the size of _2^32 - 1_ (assuming only IPv4 addresses), which will be a massive waste of memory given that we only have a small number of servers to monitor (certainly much less than 4.29 billion!). Instead, we will divide the IP range into _K_ even segments to be our partitions. We can then convert the string IP address in each record from the form of `A.B.C.D` to an integer, and determine the index of the segment it is in, which will become its partition identifier:
+
+```python
+class ServerPartitionFunction(object):
+    def __init__(self, partitions):
+        self.part_size = 2**32/partitions
+
+    def partition(self, data):
+        parts = data['server'].split('.')
+        ipnum = ((int(parts[0]) << 24) + (int(parts[1]) << 16) +
+                 (int(parts[2]) << 8) + int(parts[3]))
+        return ipnum/self.part_size
+```
+
+The state object is where things get interesting. For each server, we need to maintain a 60 seconds long rolling window of aggregate data. This data is small enough that we could probably just use a list and pay the cost of shifting it whenever we need to roll the window. But a more efficient way of doing this, especially if the data window is larger, would be to use a [circular buffer]. For brevity, the code for the circular buffer is not included in this post, but you may find in the [demo code][ddos-detection circularbuffer].
+
+Each server's state needs to keep track of a few things:
+
+1. `address` - the server's address
+2. `window` - a rolling window of per-second data aggregates for requests per second and unique clients per second
+3. `requests_mean` - the weighted mean of the requests data over the window
+4. `requests_stdev` - the weighted standard deviation of the requests data over the window
+5. `clients_mean` - the weighted mean of the clients data over the window
+6. `clients_stdev` - the weighted standard deviation of the clients data over the window
+7. `current_ts` - the current partial second's timestamp
+8. `current_requests` - the number of requests for the current (partial) second
+9. `current_clients` - a set of unique client addresses (to produce the number of clients per second aggregate)
+10. `is_attack` - the current health status of the server
+
+```python
+class SingleServerLoadStatistics(object):
+    def __init__(self, address):
+        self.address = address
+        self.window = CircularBuffer(60)
+        self.requests_mean = 0
+        self.requests_stdev = 0
+        self.clients_mean = 0
+        self.clients_stdev = 0
+        self.current_ts = 0
+        self.current_requests = 0
+        self.current_clients = set()
+        self.is_attack = False
+```
+
+That seems simple enough. But we also need to implement three pieces of logic:
```

Comment:
I don't think there is any value in "That seems simple enough". For someone who finds it simple, its throw away text. For someone who doesn't find it simple, it's frustrating. I think this could be:

"We also need to implement three pieces of logic:"